app/src/lib/helpers/regex.ts

```diff
@@ -0,0 +1,41 @@
+/**
+ * get all regex captures within a body of text
+ * @param text string to search
+ * @param re regex to search with. must have global option and one capture
+ * @returns set of strings captured by supplied regex
+ */
+export async function getCaptures(
+  text: string,
+  re: RegExp
+): Promise<Set<Array<string>>> {
+  const matches = await getMatches(text, re)
+  const captures = matches.reduce(
+    (captures, match) => captures.add(match.slice(1)),
+    new Set<Array<string>>()
+  )
+  return captures
+}
+
+/**
+ * get all regex matches within a body of text
+ * @param text string to search
+ * @param re regex to search with. must have global option
+ * @returns set of strings captured by supplied regex
+ */
+export async function getMatches(
```

Comment:
@iAmWillShepherd does this look better? using promises should keep us from overflowing the stack and i've moved the return value from the function arguments.

i'm trying to avoid a `while` loop since that would be blocking

Reply:
Not really a fan of using recursion in general when iteration is typically much faster. If JS has proper tail recursion, this would be a non-issue for me. But at the moment, it seems like we trade the possibility of blowing the stack for having thousands of promises queued up.

Reply:
in my testing of this function, the promise chain is faster (and doesn't block the process) in comparison to a while loop. on my machine its about 20% faster

this won't create thousands of promises at once because they are created recursively in a chain. aka each promise is only created at the end of the preceding promise.